File: DSA/linkedlist/linkedlist.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
    int data;
    struct node *next;
} node;

// typedef struct linkedList {
//     struct node *head;
//     static node *tail;
// } linkedList

node *createNode(int data) {
    node *newNode = (node *)malloc(sizeof(node));
    newNode->data = data;
    newNode->next = NULL;

    return newNode
}

node *insertNode(node *head, int data){
    if (head == NULL) {
        head = createNode(data);
    } else {
        root->next = insertNode(root->next, data);
    }
    return head
}

node *removeNode(node *head, int data) {
    if (head == NULL) {
        return;
    } else {
        if (head->data == data) {
            head = head->next;
        } else {
            head->next = removeNode(head->next, data);
        }
    }

    return head;
}


int searchNode(node *head, int data) {
    if (head == NULL) {
        return 0;
    } else if (head->data == data) {
        return 1;
    } else {
        return searchNode(head->next, data);
    }
}

void printList(node *head) {
    if (head == NULL) {
        return;
    }

    printf("%d->", head->data);
    printList(head->next);
}

int main() {
    node *head = NULL;
    head = insertNode(head, 5);
    head = insertNode(head, 4);
    head = insertNode(head, 3);
    head = insertNode(head, 2);
    head = insertNode(head, 1);

    printList(head);
}

File: DSA/linkedlist/linkedlist.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add(self, value):
        """Add a new node to the end of the list"""
        if self.head is None:
            self.head = Node(value)
            self.tail = self.head
        else:
            self.tail.next = Node(value)
            self.tail = self.tail.next
    
    def remove(self, value): 
        if self.head is None:
            return None
        else:
            if self.head.value == value:
                self.head = self.head.next
            else:
                current = self.head
                while current.next is not None:
                    if current.next.value == value:
                        current.next = current.next.next
                        break
                    current = current.next
    
    def traverse(self, head):
        if head is None: 
            return None
        print(f"{head.value} ->")
        self.traverse(head.next)



    def __iter__(self):
        """Iterate over the list"""
        current = self.head
        while current is not None:
            yield current.value
            current = current.next

if __name__ == "__main__":
    List = LinkedList()
    List.add(1)
    List.add(2)
    List.add(3)
    List.add(4)

    List.traverse(List.head)

    List.remove(3)

    List.traverse(List.head)